bigtable-dataflow-import/src/main/java/com/google/cloud/bigtable/dataflowimport/HBaseResultToMutationFn.java

```diff
@@ -48,25 +81,78 @@ static void setLogger(Logger log) {
   @Override
   public void processElement(ProcessContext context) throws Exception {
     KV<ImmutableBytesWritable, Result> kv = context.element();
-    Cell[] cells = checkEmptyRow(kv);
-    if (cells.length == 0) {
+    List<Cell> cells = checkEmptyRow(kv);
+    if (cells.isEmpty()) {
       return;
     }
-    // TODO: Handle DeleteMarkers properly.
-    Put put = new Put(kv.getKey().get());
+    Put put = tryProcessRowWithNoDeleteMarkers(kv.getKey().get(), cells);
+    if (put == null) {
+      put = processRowWithDeleteMarkers(kv.getKey().get(), cells);
+    }
+    context.output(put);
+  }
+
+  // Optimistically process the row assuming no delete markers exist. Return null if delete markers
+  // are found.
+  private Put tryProcessRowWithNoDeleteMarkers(byte[] rowKey, List<Cell> cells) throws IOException {
+    Put put = new Put(rowKey);
     for (Cell cell : cells) {
+      if (CellUtil.isDelete(cell)) {
+        // Delete Marker found: abort and let caller invoke processRowWithDeleteMarkers().
+        return null;
+      }
       put.add(cell);
     }
-    context.output(put);
+    return put;
+  }
+
+  // Process
+  private Put processRowWithDeleteMarkers(byte[] rowKey, List<Cell> cells) throws IOException {
+    Put put = new Put(rowKey);
+    // Group cells by column family, since DeleteMarkers do not apply across families.
+    Map<String, Collection<Cell>> dataCellsByFamilyMap =
+        Multimaps.index(
+            Iterables.filter(cells, Predicates.not(IS_DELETE_MARKER_FILTER)),
+            COLUMN_FAMILY_EXTRACTOR).asMap();
+    Map<String, Collection<Cell>> deleteMarkersByFamilyMap =
+        Multimaps.index(
+            Iterables.filter(cells, IS_DELETE_MARKER_FILTER), COLUMN_FAMILY_EXTRACTOR).asMap();
```

Comment:
Here we are making an assumption that if there is a delete marker for this row, it has to be here in "cells". Is it always true as result of reading sequence file?